File: erlang/kmeans.erl
-module(kmeans).
-compile(export_all).

assign(Means, Data, Chunksize) ->
  L = length(Data),
  if
    L =< Chunksize ->
      assign_sequential(Means, Data);
    true ->
      {Part, Rest} = lists:split(Chunksize, Data),
      Parent = self(),
      R = make_ref(),
      spawn_link(fun() ->
        Parent ! {R, assign_sequential(Means, Part)}
      end),
      RestRes = assign(Means, Rest, Chunksize),
      PartRes = receive {R, Res} -> Res end,
      lists:append(PartRes, RestRes)
  end.

assign_sequential(Means, Data) ->
  lists:map(fun(Point) ->
    closest(lists:map(fun(Mean) -> distance(Point, Mean) end, Means))
  end, Data).

closest(Ds) ->
  element(2, lists:min(
    lists:zip(Ds, lists:seq(0, length(Ds) - 1)))
  ).

distance({Px, Py}, {Qx, Qy}) ->
  math:pow(Px - Qx, 2) + math:pow(Py - Qy, 2).

File: haskell/kmeans.hs
module Kmeans where

import Control.Monad.Par

data Point = Point {
  x :: Double,
  y :: Double
}

square :: Double -> Double
square a = a * a

distance :: Point -> Point -> Double
distance p q = square (x p - x q) + square (y p - y q)

closest :: [Double] -> Int
closest xs = snd . minimum $ zip xs [0..length xs]

assign :: [Point] -> [Point] -> [Int]
assign means = map (\p -> closest (map (`distance` p) means))

passign :: [Point] -> [Point] -> Int -> Par [Int]
passign means ps chunksize | length ps <= chunksize = return $ assign means ps
                           | otherwise              = do
                               let (part, rest) = splitAt chunksize ps
                               ipart <- spawnP (assign means part)
                               crest <- passign means rest chunksize
                               cpart <- get ipart
                               return $ cpart ++ crest
